### PR TITLE
changing parallel loops to Future.apply

### PR DESCRIPTION
**Pull request by parismita**
issue #32

### CODE REVIEW

File: Rpackage/DESCRIPTION [hunk 1]
```diff
@@ -6,6 +6,6 @@ License: GPL-3
 Title: Max Margin Interval Trees
 Description: Fast O(P N log N) algorithm for learning a regression tree
  with interval censored output data.
-Depends: partykit, penaltyLearning, assertthat, doParallel, parallel, foreach
+Depends: partykit, penaltyLearning, assertthat, parallel, doParallel, future.apply
```
Comment (tdhock) [inline Rpackage/DESCRIPTION]: remove doParallel since that is not used anymore right?

Change Depends to Imports please! (and update the NAMESPACE accordingly)

Depends means that when you do library(mmit) it will do library(pkg) for each of the Depends packages. (which attaches each Depends package, making all of their exported functions available for the user)

Typically Imports is preferred because when the user types library(mmit) they usually want access to mmit functions but probably not all of partkit, penaltyLearning, assertthat, etc functions.

Comment (parismita) [reply Rpackage/DESCRIPTION]: ok

**Comment by parismita** [comment]
@tdhock I converted to future.apply but to use future::plan(), I need to import future, then it gives the following warning:
```
Warning message:
replacing previous import ‘future::future_lapply’ by ‘future.apply::future_lapply’ when loading ‘mmit’ 
```
But if I use future:future_lapply, that function is deprecated so it gives warning again.

What shall I do now?

**Comment by tdhock** [comment]
hi you should not use future::plan in your package

the user of your package uses future::plan to tell the system how to
implement the parallel computation

**Comment by parismita** [comment]
ok, so the user will define the no of cpus as well as the implementation.....So shall I remove it?

**Comment by tdhock** [comment]
yes you should remove future::plan in your package code

but you need to keep the future.apply::future_lapply in your package.

**Comment by parismita** [comment]
Will the user use makeCluster by itself?

**Comment by tdhock** [comment]
the user gets to decide how to do the parallelism, maybe the user will use
parallel package / makeCluster, maybe the user will use something else...

**Comment by parismita** [comment]
ok...I'll remove them and add a note in the documentation regarding the same...that the user gets to decide how to do the parallelism

**Comment by tdhock** [comment]
great.

for an example of how I did it in another package see
https://github.com/tdhock/penaltyLearning/blob/master/R/IntervalRegression.R#L99

a good idiom is

LAPPLY <- if(requireNamespace("future.apply")){
future.apply::future_lapply
}else{
lapply
}

so then future.apply is not a hard dependency (you should put it under
Suggests in DESCRIPTION)

**Comment by parismita** [comment]
Hi @tdhock , when I check using --as-cran, it gives a note 
```
checking compiled code ... NOTE
File ‘mmit/libs/mmit.so’:
  Found ‘_ZSt4cout’, possibly from ‘std::cout’ (C++)
    Object: ‘piecewise_function.o’
File ‘mmit/libs/mmit.so’:
  Found no calls to: ‘R_registerRoutines’, ‘R_useDynamicSymbols’

Compiled code should not call entry points which might terminate R nor
write to stdout/stderr instead of to the console, nor use Fortran I/O
nor system RNGs.
It is good practice to register native routines and to disable symbol
search.
```

I tried to change the std::cout to Rcpp:Rcout, but it throws error....what should be done in this case?

**Comment by tdhock** [comment]
well you probably should just delete the cout (usually C++ code should be
silent if you are not debugging it)

about the register routines, see
https://www.ironholds.org/registering-routines/

File: .Rd2pdf12385/Rd2.tex [hunk 2]
```diff
@@ -0,0 +1,12 @@
+\documentclass[letterpaper]{book}
+\usepackage[times,inconsolata,hyper]{Rd}
+\usepackage{makeidx}
```
Comment (tdhock) [inline .Rd2pdf12385/Rd2.tex]: this file should not be in the git repo

Comment (parismita) [reply .Rd2pdf12385/Rd2.tex]: ok

**Comment by parismita** [comment]
@tdhock I didnt understand the error that occured...is it because of the .c file?

**Comment by tdhock** [comment]
all builds fail because mvtnorm is not available for R 3.4.4 https://travis-ci.org/aldro61/mmit/jobs/508549725#L2560

that is a relatively old version of R

try upgrading the version of R used in .travis.yml

**Comment by aldro61** [comment]
@parismita Do you want to merge master into this branch to fix the R version problems? 

Also, as discussed by email, the adaboost code got merged here by mistake. Since the mmitboost.R and mmitboost.predict.R files are not completely ready, we will delete them from here. However, we will keep the modified solver with example weights, since I already approved this code in another PR. Does this make sense?

**Comment by parismita** [comment]
@aldro61 Yeah, i'll pull from the master to update the travis part... 

I'll delete the mmitboost part and make a new branch for adaboost for that

Thanks :)

**Comment by parismita** [comment]
@aldro61 Yeah, this branch is already upto date with the latest R version... @tdhock I completed the cran test without any warning/note/error 

I'll delete the mmitboost part and make a new branch for adaboost for that

Thanks :)

**Comment by parismita** [comment]
@tdhock @aldro61 I updated it to make it cran submitable and deleted the adaboost R codes

File: Rpackage/R/mmif.R [hunk 3]
```diff
@@ -30,58 +29,46 @@
...
   
 }, ex=function(){
   
-  data(neuroblastomaProcessed, package="penaltyLearning")
+  data("neuroblastomaProcessed", package="penaltyLearning", envir=environment())
   feature.mat <- data.frame(neuroblastomaProcessed$feature.mat)[1:45,]
   target.mat <- neuroblastomaProcessed$target.mat[1:45,]
+  if(require(parallel)){
+    cl <- makeCluster(detectCores())
+    registerDoParallel(cl)
```
Comment (tdhock) [inline Rpackage/R/mmif.R]: Shouldn't this create a WARNING or ERROR because registerDoParallel is in the foreach package, which is no longer imported?

Maybe the Rd files need to be re-generated via `inlinedocs::package.skeleton.dx()` ?

anyways please delete this if statement since foreach is no longer used

File: Rpackage/R/mmif.R [hunk 4]
```diff
@@ -30,58 +29,46 @@
...
+  }
+  if(require(future)){ plan(multiprocess)}
   trees <- mmif(target.mat, feature.mat, max_depth = Inf, margin = 2.0, loss = "hinge", min_sample = 1)
-  
+  if(require(parallel)){ stopCluster(cl)}
```
Comment (tdhock) [inline Rpackage/R/mmif.R]: also please delete this

File: Rpackage/R/mmif.cv.R [hunk 5]
```diff
@@ -104,7 +98,13 @@ mmif.cv <- structure(function(target.mat, feature.mat,
   param_grid$n_trees <- c(10)
   param_grid$n_features <- c(as.integer(ncol(feature.mat)**0.5))
   
-  result <- mmif.cv(target.mat, feature.mat, param_grid, scorer = mse, n_cpu = 4)
+  if(require(parallel)){
+    cl <- makeCluster(detectCores())
+    registerDoParallel(cl)
```
Comment (tdhock) [inline Rpackage/R/mmif.cv.R]: please delete

File: Rpackage/R/mmif.cv.R [hunk 6]
```diff
@@ -104,7 +98,13 @@ mmif.cv <- structure(function(target.mat, feature.mat,
...
+  }
+  if(require(future)){ plan(multiprocess)}
+  result <- mmif.cv(target.mat, feature.mat, param_grid, scorer = mse)
+  if(require(parallel)){ stopCluster(cl)}
```
Comment (tdhock) [inline Rpackage/R/mmif.cv.R]: please delete

File: Rpackage/R/mmit.cv.R [hunk 7]
```diff
@@ -99,6 +91,11 @@ mmit.cv <- structure(function(target.mat, feature.mat,
   param_grid$min_sample <- c(2, 5, 10, 20)
   param_grid$loss <- c("hinge", "square")
   
-  result <- mmit.cv(target.mat, feature.mat, param_grid, scorer = mse, n_cpu = 4)
-  
+  if(require(parallel)){
+    cl <- makeCluster(detectCores())
+    registerDoParallel(cl)
```
Comment (tdhock) [inline Rpackage/R/mmit.cv.R]: please delete

File: Rpackage/R/mmit.cv.R [hunk 8]
```diff
@@ -99,6 +91,11 @@ mmit.cv <- structure(function(target.mat, feature.mat,
...
+  }
+  if(require(future)){ plan(multiprocess)}
+  result <- mmit.cv(target.mat, feature.mat, param_grid, scorer = mse)
+  if(require(parallel)){ stopCluster(cl)}
```
Comment (tdhock) [inline Rpackage/R/mmit.cv.R]: please delete

File: Rpackage/src/registerDynamicSymbol.c [hunk 9]
```diff
@@ -0,0 +1,10 @@
+// RegisteringDynamic Symbols
+
+#include <R.h>
+#include <Rinternals.h>
+#include <R_ext/Rdynload.h>
+
+void R_init_markovchain(DllInfo* info) {
```
Comment (tdhock) [inline Rpackage/src/registerDynamicSymbol.c]: should be R_init_mmit

File: tutorials/R/README.md [hunk 10]
```diff
@@ -306,10 +299,12 @@ The collection of trees are :
 ![alt text](./Selection_017.png "Random Forest")
 ![alt text](./Selection_018.png "Random Forest")
 
+*<b> Warning </b> : The model and predictions may vary from the example due to random sampling*
```
Comment (tdhock) [inline tutorials/R/README.md:302]: would be better to use set.seed(1) so that model/predictions are the same

File: tutorials/R/README.md [hunk 11]
```diff
@@ -339,6 +334,8 @@ print(pred)
 
 pred : 0.75 0.95 1.95 2.15 2.15 2.15
 
+*<b> Warning </b> : The model and predictions may vary from the example due to random sampling*
```
Comment (tdhock) [inline tutorials/R/README.md:337]: set.seed

File: tutorials/R/README.md [hunk 12]
```diff
@@ -381,7 +371,9 @@ Best parameters :
 
 max_depth | margin |min_sample | loss |n_trees |n_features|
 | ---- |:------: | ---------:|---: |---: |---: |
-3         |5       |          2|  hinge|      10|          2|
+3         |2       |          5|  square |      10|          10|
+
+*<b> Warning </b> : The model and predictions may vary from the example due to random sampling*
```
Comment (tdhock) [inline tutorials/R/README.md:376]: also it would be better to put this tutorial in the standard R pkg location, e.g. mmit/Rpackage/vignettes/tutorial.Rmd

see http://kbroman.org/pkg_primer/pages/vignettes.html for more info about how to write vignettes. it is ok to leave these warnings as is for this PR, but please open a separate PR for moving this tutorial to a vignette.

File: Rpackage/R/mmif.R (hunk 3 above)
Comment (parismita) [reply Rpackage/R/mmif.R]: ok

File: tutorials/R/README.md (hunk 10 above)
Comment (parismita) [reply tutorials/R/README.md:302]: ok...I'll add the seed and generate the trees again

File: tutorials/R/README.md (hunk 12 above)
Comment (parismita) [reply tutorials/R/README.md:376]: ok....and should I put the seed in random forest code in this PR and update the doc or open new PR for that?

File: Rpackage/R/mmit.cv.R (hunk 8 above)
Comment (tdhock) [reply Rpackage/R/mmit.cv.R]: actually all of the ex=function(){} blocks were example code that was converted to Rd examples by inlinedocs. but since you are using roxygen2 now, it does not make sense to include these anymore. can you please either (1) delete them, or (2) move them into the roxygen2 examples?

File: tutorials/R/README.md (hunk 12 above)
Comment (tdhock) [reply tutorials/R/README.md:376]: please keep this PR focused on the stated purpose (changing parallel for loops to future.apply) and open a new PR for the random seed/vignette

Comment (parismita) [reply tutorials/R/README.md:376]: ok

File: Rpackage/R/mmit.cv.R (hunk 8 above)
Comment (parismita) [reply Rpackage/R/mmit.cv.R]: ok...i have already given roxygen2 examples so I guess i'll remove these ones....I was using them for debugging purpose...but in final package they shouldnt be there i guess...i'll remove them

**Comment by parismita** [comment]
@tdhock I didnt understand why the travis failed...what might have gone wrong?

File: Rpackage/R/mmit.cv.R (hunk 8 above)
Comment (parismita) [reply Rpackage/R/mmit.cv.R]: done in vignettes pull request

**Comment by tdhock** [comment]
probably some network issue, in that case just click on re-run build (I already did)

**Comment by parismita** [comment]
@tdhock i think its because the branch is showing to be vignettes....it is supposed to be master

**Comment by tdhock** [comment]
https://stackoverflow.com/questions/39140800/the-command-eval-git-fetch-origin-refs-pull-7-merge-failed

**Comment by parismita** [comment]
@tdhock yeah...i updated the vignettes branch by merging from this branch...maybe the travis build was not completed then so its giving error.....but I am unable to revert the merge as I did commits after the merge...so currently I can rewrite the content of this branch to a new branch and have a pull req

**Comment by tdhock** [comment]
i would suggest avoiding re-writing the git history if at all possible.

the vignettes PR https://github.com/aldro61/mmit/pull/35 seems like it does NOT have the commits from this branch.

maybe try opening a new PR for this branch?

**Comment by parismita** [comment]
yeah...i'll be more carefull regarding pr, merges and git history next time...I'll try to open a new pr for this for now

**Comment by parismita** [comment]
@tdhock  even this PR's build is clean...i dont understand how

**Review by tdhock — APPROVED** [review]
not sure why the tests are working now, but it looks ok, please merge

**Comment by parismita** [comment]
ok
